nixos/modules/services/web-apps/nextcloud.nix

```diff
@@ -85,18 +85,56 @@ let
     "-dmemory_limit=${cfg.cli.memoryLimit}"
   ]);
 
-  occ = pkgs.writeScriptBin "nextcloud-occ" ''
-    #! ${pkgs.runtimeShell}
-    cd ${webroot}
-    sudo=exec
-    if [[ "$USER" != nextcloud ]]; then
-      sudo='exec /run/wrappers/bin/sudo -u nextcloud'
-    fi
-    $sudo ${pkgs.coreutils}/bin/env \
-      NEXTCLOUD_CONFIG_DIR="${datadir}/config" \
-      ${phpCli} \
-      occ "$@"
-  '';
+  # NOTE: The credentials required by all services at runtime, not including things like the
+  #       admin password which is only needed by the setup service.
+  runtimeSystemdCredentials = []
+    ++ (lib.optional (cfg.config.dbpassFile != null) "dbpass:${cfg.config.dbpassFile}")
+    ++ (lib.optional (cfg.config.objectstore.s3.enable) "s3_secret:${cfg.config.objectstore.s3.secretFile}")
+    ++ (lib.optional (cfg.config.objectstore.s3.sseCKeyFile != null) "s3_sse_c_key:${cfg.config.objectstore.s3.sseCKeyFile}");
+
+  requiresRuntimeSystemdCredentials = (lib.length runtimeSystemdCredentials) != 0;
+
+  occ = pkgs.writeShellApplication {
+    name = "nextcloud-occ";
+
+    text = let
+      command = ''
+        ${lib.getExe' pkgs.coreutils "env"} \
+          NEXTCLOUD_CONFIG_DIR="${datadir}/config" \
+          ${phpCli} \
+          occ "$@"
+      '';
+    in ''
+      cd ${webroot}
+
+      # NOTE: This is templated at eval time
+      requiresRuntimeSystemdCredentials=${lib.boolToString requiresRuntimeSystemdCredentials}
+
+      # NOTE: This wrapper is both used in the internal nextcloud service units
+      #       and by users outside a service context for administration. As such,
+      #       when there's an existing CREDENTIALS_DIRECTORY, we inherit it for use
+      #       in the nix_read_secret() php function.
+      #       When there's no CREDENTIALS_DIRECTORY we try to use systemd-run to
+      #       load the credentials just as in a service unit.
+      # NOTE: If there are no credentials that are required at runtime then there's no need
+      #       to load any credentials.
```

Comment:
IMO we should always execute the script the same way to avoid edge cases and to make testing changes easier.

Reply:
If there are no credentials the script would try to use `systemd-run` from within other services, which fails.

It needs some way of knowing if its running in a service or that it shouldn't expect credentials

Reply:
On a first glance I'd be fine with that, but it would be nice to have a test which runs `occ` from the "outside" to make sure we immediately notice if we're about to break it (that means, a command that fails if secrets are unreadable).

Reply:
We have very basic checks for `nextcloud-occ` everywhere: https://github.com/NixOS/nixpkgs/blob/5f81eeb3e39007bcdc3c17488b3225adf2102c84/nixos/tests/nextcloud/default.nix#L70

If you have ideas to expand this let me know

Reply:
This is the output of `nextcloud-occ status` with logging enabled for whenever a file path is being read

```diff
       function nix_read_secret($file) {
         $credential_file = getenv("CREDENTIALS_DIRECTORY") . "/" . $file;
+        error_log($credential_file);
         if (!is_readable($credential_file)) {
```

```
$ nextcloud-occ status
/run/credentials/run-u13.service/dbpass
  - installed: true
  - version: 30.0.4.1
  - versionstring: 30.0.4
  - edition: 
  - maintenance: false
  - needsDbUpgrade: false
  - productname: Nextcloud
  - extendedSupport: false
```

I've just pushed additional changes that make the script exit with error code 1, the previously thrown `RuntimeException` would be caught somewhere:

```
subtest: Ensure nextcloud-occ is working
nextcloud: must succeed: nextcloud-occ status
nextcloud # [   23.857755] systemd[1]: Starting /nix/store/b1wvkjx96i3s7wblz38ya0zr8i93zbc5-coreutils-9.5/bin/env NEXTCLOUD_CONFIG_DIR=/var/lib/nextcloud/config /nix/store/c6i258wgrwgg8aladc0v97k39cs2sgp6-php-with-extensions-8.3.15/bin/php occ status...
nextcloud # [   23.897224] systemd[1]: Started /nix/store/b1wvkjx96i3s7wblz38ya0zr8i93zbc5-coreutils-9.5/bin/env NEXTCLOUD_CONFIG_DIR=/var/lib/nextcloud/config /nix/store/c6i258wgrwgg8aladc0v97k39cs2sgp6-php-with-extensions-8.3.15/bin/php occ status.
nextcloud # [   23.984784] nextcloud-update-db-start[1148]: oc_filecache table updated successfully.
nextcloud # [   23.985753] nextcloud-update-db-start[1148]: Adding additional systag_by_objectid index to the oc_systemtag_object_mapping table, this can take some time...
nextcloud # [   24.110756] systemd[1]: run-u4.service: Main process exited, code=exited, status=1/FAILURE
nextcloud # [   24.112105] systemd[1]: run-u4.service: Failed with result 'exit-code'.
nextcloud: output: Cannot read credential 'dbpass' passed by NixOS, $CREDENTIALS_DIRECTORY is not set!

Test "Ensure nextcloud-occ is working" failed with error: "command `nextcloud-occ status` failed (exit code 1)"
```

Reply:
Looks reasonable, thanks!